src/shell.ts

```diff
@@ -0,0 +1,170 @@
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async execute(cmd: string, args: string[]) {
+        if (this.activeForm) 
+            this.escapeActiveForm()
+        if (!cmd)
+            return
+        const command = this.commands.get(cmd)
+        if (!command)
+            return this.t.writeln(`Command not found: ${cmd}`)
+        this.active = false
+        await command.execute(args)
+        this.active = true
+    }
+
+    async runCommand(cmd: string, args: string[]) {
+        this.escapeActiveForm()
+        this.map.interruptTTY()
+        this.field = [cmd, ...args].join(' ')
+        this.printPrompt()
+        this.t.write("\n")
+        await this.execute(cmd, args)
+        this.field = ''
+        this.printPrompt()
+    }
+
+    async newForm(fields: Fields, type: "menu" | "choice" | "text", title="") {
+        this.escapeActiveForm()
+        this.map.showLog(true)
+        this.t.write("\r\x1B[K")
+        this.t.scrollToBottom()
+        if (title)
+            this.t.writeln(title)
+        this.activeForm = new Form(fields)
+        let promise
+        switch (type) {
+            case "menu":
+                promise = this.activeForm.menu(this.t)
+                break
+            case "choice":
+                promise = this.activeForm.chooseFields(this.t)
+                break
+            case "text":
+                promise = this.activeForm.start(this.t)
+                break
+            default:
+                throw new Error("Unknown form type: " + type)
+        }
+        try {
+            const res = await promise
+            return res
+        } catch (err) {
+            this.onFormError(err)
+            throw err
+        } finally {
+            this.activeForm = null
+            this.printPrompt()
+        }
+    }
+
+    escapeActiveForm() {
+        if (!this.activeForm) return
+        this.t.scrollToBottom()
+        this.printBelowForm("ESC\n")
+        this.activeForm.reject(new Error("aborted"))
+        this.activeForm = null
+    }
+    
+    keyHandler(ev: KeyboardEvent) {
+        const form = this.activeForm,
+            key = ev.key
+        this.printPrompt()
+        if (key == 'Escape') {
+            if (form)
+                this.escapeActiveForm()
+            else
+                this.map.showLog(false)
+        } else if ((ev.ctrlKey || ev.metaKey) && (key == 'v')) {
+            Clipboard.read().then(res => {
+                if (res.type == 'text/plain') {
+                    form.field += res.value
+                    if (!form.hidden)
+                        this.t.write(res.value)
+                }
+            })
+        } else if (form?.onKey)
+            form.onKey(ev)
+        else if (this.active)
+            this.onKey(ev)
+        ev.preventDefault()
+    }
+
+    onFormError(err: Error) {
+        terminal7.log("Form error: " + err)
+        terminal7.clearTempGates()
```

Comment:
I'm not sure it belongs here. Can we catch the form reject on the caller and clear the gate there?